docs/docs/reference/metaprogramming/quote-matching.md

```diff
@@ -0,0 +1,64 @@
+---
+layout: doc-page
+title: "Quote Matching Spec"
+---
+
+## Evaluation sematics
+
+Bellow we provide an informal description of the pattern evailatuion.
+We use `'{..}` for expression, `'[..]` for types and `{{..}}` for patterns nested in expressions.
+The semantics are defined as a list of reduction rules that are tried one by one until one matches.
+
+Operations:
+* `s =?= p` checks if a scrutinee `s` matches the pattern `p` while accumulating extracted parts of the code.
+* `isColosedUnder(x1, .., xn)('{e})` returns true if and only if all the references in `e` to names defined in the patttern are contained in the set `{x1, ... xn}`.
+* `lift(x1, .., xn)('{e})` returns `(y1, ..., yn) => [xi = $yi]'{e}` where `yi` is an `Expr` of the type of `xi`.
+* `withEnv(x1 -> y1, ..., xn -> yn)(matching)` evaluates mathing recording that `xi` is equivalent to `yi`.
+* `matched` denotes that the the match succedded and `matched('{e})` denotes that a matech succeded and extracts `'{e}`
+* `&&&` matches if both sides match. Concatenates the extracted expressions of both sides.
+
+```scala
+'{ e } =?= '{ $x }  &&  typeOf('{e}) <:< typeOf('{$x}) && isColosedUnder()('{e})  ===>   matched('{e})
+'{ e } =?= '{ $y(x1, ..., xn) }  &&  isColosedUnder(x1, ... xn)('{e})  ===>   matched(lift(x1, ..., xn)('{e}))
+
+'{ lit } =?= '{ lit }   ===>   matched
+'{ e: T } =?= '{ p: P }   ===>   '{e} =?= '{p} &&& '[T] =?= '[P]
+'{ e } =?= '{ p: P }   ===>   '{e} =?= '{p}
+'{ e.x } =?= '{ p.x }   ===>   '{e} =?= '{p}
+'{ x } =?= '{ x }   ===>   matched
```

Comment:
Should this be somewhere next with `'{{ `x` }} =?= '{{ `y` }}` , and maybe merged together?

Reply:
I now suspect I agree with this rule, but it took me a while to see why `'{ x } =?= '{ y }` is illegal (because alpha-equivalence is handled when checking local binders). Some explanation would help. (Who writes it is another question).

BTW, I'd hope `Select`s encoded as `Ident`s (per `desugarIdent`) show up as `Select` here. I see they do not in Tasty reflection... does at least -Ycheck ensure that they're represented consistently so it's well-defined what you have to write to match `Int` (as opposed to needing `Int`, `scala.Int` or `_root_.scala.Int` depending on, say, what the user wrote)?